### PR TITLE
chore: update to 0.47

### LINKED ISSUE #2650: Migrate to cosmos/gogoproto
## Summary

Migrate from `gogo/protobuf` to `cosmos/gogoproto`, [ref issue](https://github.com/cosmos/cosmos-sdk/issues/10925) and [ref pr](https://github.com/cosmos/cosmos-sdk/pull/13070)

## Problem Definition

We are still using the go mod replace directive to [unmaintained code](https://github.com/regen-network/protobuf/)

## Proposal

Switch from the regen fork to the cosmos fork. I'm unsure how involved this work will be outside of updating imports

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged/assigned

### PR DESCRIPTION
**Pull request by tac0turtle**
## Description

update to 0.47

Closes #2650
Closes: #2180
Closes: #2671 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/10-structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [x] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

**Comment by tac0turtle** [comment]
tests fail cause im using a local replace

**Comment by crodriguezvega** [comment]
Thanks a lot for taking up this work, @tac0turtle. Would you mind though to target the PR to [this feature branch](https://github.com/cosmos/ibc-go/tree/upgrade-sdk-0.47) that we had? We can merge PRs to that branch until the work is completed and smaller PRs will be also easier for us to review.

**Comment by julienrbrt** [comment]
> Thanks a lot for taking up this work, @tac0turtle. Would you mind though to target the PR to [this feature branch](https://github.com/cosmos/ibc-go/tree/upgrade-sdk-0.47) that we had? We can merge PRs to that branch until the work is completed and smaller PRs will be also easier for us to review.

The diff might seem big, but there is actually not so much changes. Most of it is due to regeneration of proto files and the deletion of the `third_party` folder. And that part isn't really splittable.

**Comment by tac0turtle** [comment]
cosmos/gogo needs to be done with 0.47. it removes the replace tag in the go.mod

File: e2e/go.mod [hunk 1]
```diff
@@ -1,17 +1,18 @@
...
 
-go 1.18
+go 1.19
 
 replace github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
 
 require (
-	github.com/cosmos/cosmos-sdk v0.46.2
+	github.com/cosmos/cosmos-sdk v0.46.1
+	github.com/cosmos/gogoproto v1.4.2
```
Comment (julienrbrt) [inline e2e/go.mod:9]: We are blocked on ICS23 to upgrading to cosmos/gogoproto for fixing the last tests.

**Comment by tac0turtle** [comment]
@crodriguezvega which branch should we point this to? is there a way to get this as an alpha tag?

**Comment by crodriguezvega** [comment]
> @crodriguezvega which branch should we point this to? is there a way to get this as an alpha tag?

You can point to [this branch](https://github.com/cosmos/ibc-go/tree/upgrade-sdk-0.47). And yes, we can cut an alpha tag from that branch when this PR is merged. It would be good to get a confirmation of whether this SDK upgrade will be API breaking for ibc-go or not.

File: .github/workflows/e2e-compatibility-workflow-call.yaml [hunk 2]
```diff
@@ -2,15 +2,14 @@ on:
   workflow_call:
     inputs:
       docker-tag:
-        description: 'Docker tag being used'
+        description: "Docker tag being used"
```
Comment (colin-axner) [inline .github/workflows/e2e-compatibility-workflow-call.yaml:5]: I'd prefer not to include these changes unless they provide some benefit?

File: go.mod [hunk 3]
```diff
@@ -138,25 +146,25 @@ require (
...
 	nhooyr.io/websocket v1.8.6 // indirect
+	pgregory.net/rapid v0.5.3 // indirect
 	sigs.k8s.io/yaml v1.3.0 // indirect
 )
 
 replace (
 	// dragonberry replace for ics23
 	github.com/confio/ics23/go => github.com/cosmos/cosmos-sdk/ics23/go v0.8.0
-
 	github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
```
Comment (colin-axner) [inline go.mod:169]: Should our dependency on `gogo/protobuf` be removed?

File: modules/apps/27-interchain-accounts/module.go [hunk 4]
```diff
@@ -129,26 +129,6 @@ func (am AppModule) InitModule(ctx sdk.Context, controllerParams controllertypes
 func (AppModule) RegisterInvariants(ir sdk.InvariantRegistry) {
 }
 
-// Route implements the AppModule interface
-func (AppModule) Route() sdk.Route {
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/module.go:133]: It's nice to see these finally removed :)

File: modules/core/02-client/proposal_handler_test.go [hunk 5]
```diff
@@ -58,11 +56,11 @@ func (suite *ClientTestSuite) TestNewClientUpdateProposalHandler() {
...
-		{
-			"unsupported proposal type", func() {
-				content = distributiontypes.NewCommunityPoolSpendProposal(ibctesting.Title, ibctesting.Description, suite.chainA.SenderAccount.GetAddress(), sdk.NewCoins(sdk.NewCoin("communityfunds", sdk.NewInt(10))))
-			}, false,
-		},
+		// {
+		// 	"unsupported proposal type", func() {
+		// 		content = distributiontypes.NewCommunityPoolSpendProposal(ibctesting.Title, ibctesting.Description, suite.chainA.SenderAccount.GetAddress(), sdk.NewCoins(sdk.NewCoin("communityfunds", sdk.NewInt(10))))
+		// 	}, false,
+		// },
```
Comment (colin-axner) [inline modules/core/02-client/proposal_handler_test.go:63]: Why is this commented out?

File: modules/core/module.go [hunk 6]
```diff
@@ -117,21 +117,11 @@ func (am AppModule) RegisterInvariants(ir sdk.InvariantRegistry) {
...
 }
 
-// Route returns the message routing key for the ibc module.
-func (am AppModule) Route() sdk.Route {
-	return sdk.Route{}
-}
-
 // QuerierRoute returns the ibc module's querier route name.
 func (AppModule) QuerierRoute() string {
 	return host.QuerierRoute
```
Comment (colin-axner) [inline modules/core/module.go:122]: Is this function still needed?

File: testing/simapp/test_helpers.go [hunk 7]
```diff
@@ -210,234 +197,3 @@ func SetupWithGenesisAccounts(genAccs []authtypes.GenesisAccount, balances ...ba
 
 	return app
 }
-
```
Comment (colin-axner) [inline testing/simapp/test_helpers.go:213]: :heart: great to see lots of red

**Review by colin-axner — COMMENTED** [review]
LGTM, nice work! Thanks for doing the update for us! Very much appreciated. 

I think we will need to add functionality to run compatibility tests (e2e tests run against older versions) off of main. Seeing the e2e's run with this pr will be a good start (e2e run with two versions of main)

File: e2e/go.mod (hunk 1 above)
Comment (colin-axner) [reply e2e/go.mod:7]: For my understanding. Is the issue because of the proto registration happens at [`init`](https://github.com/cosmos/ibc-go/blob/main/modules/core/04-channel/types/channel.pb.go#L489) and now that the imports are changing, the ics23 Spec is not being registered the same [global registry](https://github.com/gogo/protobuf/blob/v1.3.2/proto/properties.go#L535)? A chain is essentially expected to use a single proto go module for all its types

Comment (colin-axner) [reply e2e/go.mod:7]: It looks like cosmos/ics23 still needs this change. Will you handle it or should we?

File: go.mod (hunk 3 above)
Comment (tac0turtle) [reply go.mod:169]: yes this replace can be removed

File: modules/core/module.go (hunk 6 above)
Comment (tac0turtle) [reply modules/core/module.go:122]: we can remove. Haven't had a moment this week to get back to this pr

File: .github/workflows/e2e-compatibility-workflow-call.yaml (hunk 2 above)
Comment (julienrbrt) [reply .github/workflows/e2e-compatibility-workflow-call.yaml:5]: Sorry that is my linter on save.

File: e2e/go.mod (hunk 1 above)
Comment (julienrbrt) [reply e2e/go.mod:7]: Now looking at the repo it's only a question of tagging a 0.10.

File: proto/buf.yaml [hunk 8]
```diff
@@ -2,10 +2,10 @@ version: v1
...
 deps:
   # see: (https://github.com/cosmos/cosmos-sdk/tree/main/proto#sdk-x-buf)
-  - buf.build/cosmos/cosmos-sdk:8cb30a2c4de74dc9bd8d260b1e75e176
-  - buf.build/cosmos/cosmos-proto:1935555c206d4afb9e94615dfd0fad31
-  - buf.build/cosmos/gogo-proto:bee5511075b7499da6178d9e4aaa628b
-  - buf.build/googleapis/googleapis:783e4b5374fa488ab068d08af9658438
+   - buf.build/cosmos/cosmos-sdk # TODO to pin
+   - buf.build/cosmos/cosmos-proto:1935555c206d4afb9e94615dfd0fad31
+   - buf.build/cosmos/gogo-proto
+   - buf.build/googleapis/googleapis:8d7204855ec14631a499bd7393ce1970
```
[the remark this answers is not in the document]
Comment (colin-axner) [reply proto/buf.yaml:8]: We will need to update imports as well, `github.com/confio/ics23/go` -> `github.com/cosmos/ics23/go`

File: testing/chain.go [hunk 9]
```diff
@@ -317,7 +318,7 @@ func (chain *TestChain) SendMsgs(msgs ...sdk.Msg) (*sdk.Result, error) {
 	// ensure the chain has the latest time
 	chain.Coordinator.UpdateTimeForChain(chain)
 
-	_, r, err := simapp.SignAndDeliver(
+	_, r, err := simtestutil.SignCheckDeliver(
```
Comment (colin-axner) [inline testing/chain.go:321]: this needs to be undone. When we deliver a tx, we need the ibc testing package to call end block and commit(). This is because we need to track information about the resulting validator set changes. Note how [here](https://github.com/cosmos/ibc-go/blob/main/testing/chain.go#L289) we use the `res` of the end block to update the tracked validator set. I'm sure there are ways we could adjust the code on either end to make the reuse work, but I think better to not spend too much time on that now

**Comment by colin-axner** [comment]
Update to the latest commit of ics23. There are still various failing tests that each need to be investigated

File: proto/buf.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (damiannolan) [reply proto/buf.yaml:8]: Awesome, thanks for pushing this PR forward!

File: testing/simapp/app.go [hunk 10]
```diff
@@ -319,40 +326,45 @@ func NewSimApp(
...
-		appCodec, keys[authtypes.StoreKey], app.GetSubspace(authtypes.ModuleName), authtypes.ProtoBaseAccount, maccPerms, sdk.GetConfig().GetBech32AccountAddrPrefix(),
-	)
+	app.AccountKeeper = authkeeper.NewAccountKeeper(appCodec, keys[authtypes.StoreKey], authtypes.ProtoBaseAccount, maccPerms, sdk.Bech32MainPrefix, authtypes.NewModuleAddress(govtypes.ModuleName).String())
+
 	app.BankKeeper = bankkeeper.NewBaseKeeper(
-		appCodec, keys[banktypes.StoreKey], app.AccountKeeper, app.GetSubspace(banktypes.ModuleName), app.ModuleAccountAddrs(),
+		appCodec,
+		keys[banktypes.StoreKey],
+		app.AccountKeeper,
+		BlockedAddresses(),
```
Comment (crodriguezvega) [inline testing/simapp/app.go:335]: Is there a reason why we wouldn't be able to use `app.ModuleAccountAddrs()` here instead of the new function `BlockedAddresses()`? Because in `app.ModuleAccountAddrs` we take out the `ibcmock` module from the list of blocked addresses, but since that is not done in `BlockedAddresses`, I think that's causing [this test to fail](https://github.com/cosmos/ibc-go/actions/runs/3674175244/jobs/6212090001).

Basically my question is: do we need to have a new function `BlockedAddresses` or can we still use the existing one `app.ModuleAccountAddrs` we had?

Comment (colin-axner) [reply testing/simapp/app.go:335]: Yes lets use a `BlockedAddresses` function. It is a hack that `ModuleAccountAddrs()` doesn't return the mock module address. We should delete the mock module from the addresses returned in `BlockedAddresses` like how it is done for gov:
```go
delete(modAccAddrs, authtypes.NewModuleAddress(ibcmock.ModuleName).String())
```

File: proto/buf.yaml [hunk 11]
```diff
@@ -2,10 +2,11 @@ version: v1
...
+   - buf.build/cosmos/ics23:b1abd8678aab07165efd453c96796a179eb3131f
```
Comment (damiannolan) [inline proto/buf.yaml:9]: We should pin to versions for sdk and gogo-proto. Is it okay to use latest tags @julienrbrt?

File: proto/buf.yaml [hunk 12]
```diff
@@ -21,4 +22,4 @@ lint:
     - PACKAGE_VERSION_SUFFIX
     - RPC_REQUEST_STANDARD_NAME
   ignore:
-    - proofs.proto # TODO: remove this when ics23 is linted and can be obtained via buf mod
\ No newline at end of file
+    - proofs.proto # TODO: remove this when ics23 is linted and can be obtained via buf mod
```
Comment (damiannolan) [inline proto/buf.yaml:25]: We can remove this now that ics23 proofs.proto has been deleted from our proto directory

File: testing/simapp/test_helpers.go [hunk 13]
```diff
@@ -31,13 +25,12 @@ import (
...
 
 	"github.com/cosmos/ibc-go/v6/testing/mock"
-	"github.com/cosmos/ibc-go/v6/testing/simapp/helpers"
 )
 
 // DefaultConsensusParams defines the default Tendermint consensus params used in
 // SimApp testing.
-var DefaultConsensusParams = &abci.ConsensusParams{
-	Block: &abci.BlockParams{
+var DefaultConsensusParams = &tmproto.ConsensusParams{
```
Comment (damiannolan) [inline testing/simapp/test_helpers.go:32]: Is it still required to pass `ConsensusParams` to `abci.RequestInitChain` now that consensus params are managed from their own sdk module?

This is used in our testing package [here](https://github.com/cosmos/ibc-go/blob/main/testing/simapp/test_helpers.go#L102) and [here](https://github.com/cosmos/ibc-go/blob/main/testing/simapp/test_helpers.go#L203)

**Review by damiannolan — APPROVED** [review]
Approving pending outstanding questions on buf and consensus params!
LGTM! Thanks to everyone for contributing to this branch ❤️

File: testing/simapp/test_helpers.go (hunk 13 above)
Comment (damiannolan) [reply testing/simapp/test_helpers.go:32]: cc. @tac0turtle

**Review by colin-axner — APPROVED** [review]
Nice! My only remaining concern is bumping the ibctest dependency, but I think this can be done in a followup as the e2e appears to work now just fine

File: proto/buf.yaml (hunk 11 above)
Comment (julienrbrt) [reply proto/buf.yaml:9]: Yes, it's best to pin so you don't have to worry about possible proto breaking changes.
If some buf dependencies need to be upgraded because of an SDK upgrade, we will mention it in our UPGRADING.md.

File: testing/simapp/test_helpers.go (hunk 13 above)
Comment (colin-axner) [reply testing/simapp/test_helpers.go:32]: Yes I believe so. The consensus params are used to modify tendermint consensus parameters. Moving consensus params to their own SDK module was changing how SDK stores the values but not modifying how SDK tells tendermint of what the values are

Comment (damiannolan) [reply testing/simapp/test_helpers.go:32]: Cool, that makes sense to me!

File: e2e/go.mod [hunk 14]
```diff
@@ -214,3 +234,9 @@ replace (
 
 // uncomment to use the local version of ibc-go, you will need to run `go mod tidy` in e2e directory.
 replace github.com/cosmos/ibc-go/v6 => ../
+
+replace (
+	github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
```
[the remark this answers is not in the document]
Comment (damiannolan) [reply e2e/go.mod:239]: I think we can remove now

[the remark this answers is not in the document]
Comment (colin-axner) [reply e2e/go.mod:239]: I tried, but ran into some import issues? I wasn't sure if it was because of the ibctest version

File: e2e/go.mod [hunk 15]
```diff
@@ -214,3 +234,9 @@ replace (
...
+	// ibctest 0.47 @ https://github.com/strangelove-ventures/ibctest/pull/337
+	github.com/strangelove-ventures/ibctest/v6 => github.com/julienrbrt/ibctest/v6 v6.0.0-20221103221804-50f6c4c4fbdb
```
[the remark this answers is not in the document]
Comment (julienrbrt) [reply e2e/go.mod:241]: yes, bumping the fork then.

File: proto/buf.yaml (hunk 11 above)
Comment (damiannolan) [reply proto/buf.yaml]: Awesome, thanks! Updated :)

File: proto/buf.yaml [hunk 16]
```diff
@@ -2,11 +2,11 @@ version: v1
 name: buf.build/cosmos/ibc
 deps:
   # see: (https://github.com/cosmos/cosmos-sdk/tree/main/proto#sdk-x-buf)
-   - buf.build/cosmos/cosmos-sdk # TODO to pin
-   - buf.build/cosmos/cosmos-proto:1935555c206d4afb9e94615dfd0fad31
-   - buf.build/cosmos/gogo-proto
-   - buf.build/googleapis/googleapis:8d7204855ec14631a499bd7393ce1970
-   - buf.build/cosmos/ics23:b1abd8678aab07165efd453c96796a179eb3131f
+  - buf.build/cosmos/cosmos-sdk:a4d885fe2ad0a999c3486fd83bb998c416fe38c8
```
Comment (julienrbrt) [inline proto/buf.yaml:5]: You shouldn't bump it to latest yet. Best to leave it to `7ea623782a4d4f11b35627913a7bc481` for now and leave the todo. When we release, we will update `https://github.com/cosmos/cosmos-sdk/tree/main/proto#sdk-x-buf`

File: e2e/go.mod (hunk 15 above)
[the remark this answers is not in the document]
Comment (colin-axner) [reply e2e/go.mod]: Looks like we need to bump interchain-account-demo repo. Lets handle it in a followup pr

File: e2e/go.mod [hunk 17]
```diff
@@ -215,3 +234,9 @@ replace (
 
 // uncomment to use the local version of ibc-go, you will need to run `go mod tidy` in e2e directory.
 replace github.com/cosmos/ibc-go/v6 => ../
+
+replace (
+	github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
+	// ibctest 0.47 @ https://github.com/strangelove-ventures/ibctest/pull/337
+	github.com/strangelove-ventures/ibctest/v6 => github.com/julienrbrt/ibctest/v6 v6.0.0-20221212124331-0930299cd224
```
Comment (charleenfei) [inline e2e/go.mod:241]: why is this replace here?

Comment (colin-axner) [reply e2e/go.mod:241]: We need a commit from ibctest which is updated to v0.47. Currently those changes are [unmerged](https://github.com/strangelove-ventures/ibctest/pull/337)

**Review by chatton — APPROVED** [review]
LGTM

**Review by crodriguezvega — APPROVED** [review]
🚀

**Comment by faddat** [comment]
hi, I think we need to rebuild the protos.

File: modules/apps/transfer/keeper/msg_server.go [hunk 18]
```diff
@@ -50,10 +50,6 @@ func (k Keeper) Transfer(goCtx context.Context, msg *types.MsgTransfer) (*types.
 			sdk.NewAttribute(types.AttributeKeyDenom, msg.Token.Denom),
 			sdk.NewAttribute(types.AttributeKeyMemo, msg.Memo),
 		),
-		sdk.NewEvent(
```
Comment (julienrbrt) [inline modules/apps/transfer/keeper/msg_server.go:53]: With the recent SDK changes, you need to re-add that.

Comment (damiannolan) [reply modules/apps/transfer/keeper/msg_server.go:53]: Yup, missed this when pulling in the latest sdk, reverting the changes atm. Thanks for the getting the updates to baseapp in so quick 🧙

**Comment by adizere** [comment]
Many thanks to @damiannolan for helping me reproduce this. What the relayer expects to find:

```
$ simd q tx --type=hash 7759B49D9751521A58B2EFCAF5FCEB0B75BF75B2DA58787B1342BCE96F18B868 --node http://localhost:27000

code: 0
codespace: ""
data: 122D0A2B2F6962632E636F72652E636C69656E742E76312E4D7367437265617465436C69656E74526573706F6E7365
events:
- attributes:
  - index: true
    key: c3BlbmRlcg==                                 <<<<<---- looks like base64 encoded keys
    value: Y29zbW9zMXk2ZDJqZzNobjRwNm41cDl5eDQzbG55dHlrNDVnc2pxNWwzYWNk
  - index: true
    key: YW1vdW50
    value: MTA2OHN0YWtl                                 <<<<<---- also for values
  type: coin_spent

```

What the relayer actually finds with the simd from this branch:


```
code: 0
codespace: ""
data: 122D0A2B2F6962632E636F72652E636C69656E742E76312E4D7367437265617465436C69656E74526573706F6E7365
events:
- attributes:
  - index: true
    key: spender                                 <<<<<---- plain
    value: cosmos1zrc4csxpy9qhmx3647sfq677rhmduc2a93rj80
  - index: true
    key: amount
    value: 878stake
  type: coin_spent
```
